languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/SessionFactoryTest.java

```diff
@@ -42,32 +49,677 @@
+        .withInMemoryMetastore()
+        .withCryptoPolicy(policy)
+        .withStaticKeyManagementService(testMasterKey)
+        .build()) {
+      ExecutorService pool = Executors.newFixedThreadPool(numThreads);
+      LongAdder tasksCompleted = new LongAdder();
+      IntStream.range(0, numTasks)
+        .parallel()
+        .forEach(i -> {
+            pool.submit(() -> {
+              try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId + i)) {
+                byte[] payload = Integer.toHexString(i).getBytes();
+                byte[] drr = session.encrypt(payload);
+                byte[] decryptedPayload = session.decrypt(drr);
+
+                assertArrayEquals(payload, decryptedPayload);
+                tasksCompleted.increment();
+              } catch (Exception e) {
+                fail(e.getMessage());
+              }
+            });
+        });
+
+      try {
+        pool.shutdown();
+        pool.awaitTermination(60, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+
+      assertEquals(numTasks, tasksCompleted.sum());
+      assertEquals(numTasks, factory.getSessionCache().estimatedSize());
+    }
+  }
+
+  @Test
+  void testSessionCacheMultiThreadedWithMaxSessionReachedSameSession() {
+    CryptoPolicy policy = BasicExpiringCryptoPolicy.newBuilder()
+        .withKeyExpirationDays(1)
+        .withRevokeCheckMinutes(30)
+        .withCanCacheSessions(true)
+        .withSessionCacheMaxSize(1)
+        .build();
+
+    try (SessionFactory factory = SessionFactory.newBuilder(testProductId, testServiceId)
+        .withInMemoryMetastore()
+        .withCryptoPolicy(policy)
+        .withStaticKeyManagementService(testMasterKey)
+        .build()) {
+      int numThreads = 100;
+      int numTasks = numThreads * 100;
+      ExecutorService pool = Executors.newFixedThreadPool(numThreads);
+      LongAdder tasksCompleted = new LongAdder();
+      IntStream.range(0, numTasks)
+        .parallel()
+        .forEach(i -> {
+            pool.submit(() -> {
+              try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId)) {
+                byte[] payload = Integer.toHexString(i).getBytes();
+                byte[] drr = session.encrypt(payload);
+                byte[] decryptedPayload = session.decrypt(drr);
+
+                assertArrayEquals(payload, decryptedPayload);
+                tasksCompleted.increment();
+              } catch (Exception e) {
+                fail(e.getMessage());
+              }
+            });
+        });
+
+      try {
+        pool.shutdown();
+        pool.awaitTermination(60, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+
+      assertEquals(numTasks, tasksCompleted.sum());
+      assertEquals(1, factory.getSessionCache().estimatedSize());
```

Comment:
Instead of verifying this, we can verify after L:557 that number of entries in cache is less than the maxCacheSize defined.

Reply:
this isn't clear to me. that shouldn't be possible. the `maxCacheSize` is 1 in this test, and the cache will already be at that size once you're in that try block that gets a `Session`